### PR TITLE
Support MCV based cardinality estimation for text columns in ORCA

### PR DESCRIPTION
**Pull request by asubramanya**
Prior to this patch, the MCVs for text columns was not being passed to ORCA.
Hence the cardinality estimation for predicates involving text was inaccurate
and led to sub optimal plans being picked. This patch allows the MCVs to be
passed in to ORCA so that it can now estimate the cardinality using MCVs equal
and not equal operators for text columns.

Co-authored-by: Abhijit Subramanya <asubramanya@pivotal.io>
Co-authored-by: Bhuvnesh Chaudhary <bchaudhary@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`

### CODE REVIEW

File: src/backend/gpopt/translate/CTranslatorUtils.cpp [hunk 1]
```diff
@@ -2750,12 +2750,8 @@ BOOL
 CTranslatorUtils::ShouldCreateStatsBucket
```
Comment (vraghavan78) [inline src/backend/gpopt/translate/CTranslatorUtils.cpp]: can we kill this function then?

Comment (asubramanya) [reply src/backend/gpopt/translate/CTranslatorUtils.cpp]: makes sense. will remove it.

File: src/backend/gpopt/translate/CTranslatorUtils.cpp [hunk 2]
```diff
@@ -2750,12 +2750,8 @@ BOOL
...
 	(
 	OID att_type_oid
-	)
+	 )
 {
-	if (att_type_oid != TEXTOID && att_type_oid != CHAROID && att_type_oid != VARCHAROID && att_type_oid != BPCHAROID)
```
Comment (chrishajas) [inline src/backend/gpopt/translate/CTranslatorUtils.cpp]: Does a regular character type need any special handling?

File: src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp [hunk 3]
```diff
@@ -2462,14 +2462,30 @@ CTranslatorScalarToDXL::ExtractLintValueFromDatum
...
+			}
+			hash = gpos::HashValue<BYTE>(arg);
+			for (ULONG ul = 1; ul < true_length; ul++)
+			{
+				hash = gpos::CombineHashes(hash, gpos::HashValue<BYTE>(&arg[ul]));
 			}
 		}
 
-		lint_value = (LINT) (hash / 4);
+		lint_value = (LINT) hash;
```
Comment (chrishajas) [inline src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp]: Why was this changed from hash/4 to hash?

File: src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp [hunk 4]
```diff
@@ -2462,14 +2462,30 @@ CTranslatorScalarToDXL::ExtractLintValueFromDatum
...
+			 */
+			if (mdid->Equals(&CMDIdGPDB::m_mdid_bpchar))
 			{
-				hash = gpos::CombineHashes(hash, gpos::HashValue<BYTE>(&bytes[ul]));
+				arg = (BYTE *)VARDATA_ANY(bytes);
+				true_length = gpdb::BpCharLen((Datum) bytes);
+			}
+			else if (mdid->Equals(&CMDIdGPDB::m_mdid_varchar) || mdid->Equals(&CMDIdGPDB::m_mdid_text))
+			{
+				arg = (BYTE *)VARDATA_ANY(bytes);
```
Comment (d) [inline src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp]: have you considered just using the `VARSIZE_ANY_EXHDR` here? An example usage will be like `length = VARSIZE_ANY_EXHDR(bytes)` or something

File: src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp [hunk 5]
```diff
@@ -2462,14 +2462,30 @@ CTranslatorScalarToDXL::ExtractLintValueFromDatum
...
+			BYTE *arg = bytes;
+			/*
+			 * For text related types, only hash the actual data and exclude the
+			 * header and padded blanks
+			 */
+			if (mdid->Equals(&CMDIdGPDB::m_mdid_bpchar))
 			{
-				hash = gpos::CombineHashes(hash, gpos::HashValue<BYTE>(&bytes[ul]));
+				arg = (BYTE *)VARDATA_ANY(bytes);
+				true_length = gpdb::BpCharLen((Datum) bytes);
```
Comment (d) [inline src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp]: `bpcharlen` returns the length in _characters_, but I believe you want length in bytes. Do we reach here when `pg_database_encoding_max_length() != 1`?

**Comment by d** [comment]
I'd suggest you consider directly using `hashbpcharextended`, `hashtextextended` and friends, which will be less error-prone than figuring out the true byte-length of the text, vis-à-vis using `gpos::HashValue<Byte>` and `gpos::CombineHashes`

File: src/backend/gpopt/translate/CTranslatorUtils.cpp (hunk 2 above)
Comment (asubramanya) [reply src/backend/gpopt/translate/CTranslatorUtils.cpp]: This does not require any special handling.

File: src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp (hunk 3 above)
Comment (asubramanya) [reply src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp]: It seems wrong to divide by 4. For E.g hash('foo') maps to 65 and hash('foobar') maps to 66. But after dividing by 4 both map to 16. Now the LINT values are no longer unique.

Comment (berlin-ab) [reply src/backend/gpopt/translate/CTranslatorScalarToDXL.cpp]: This seems like an appropriate thing to capture in a test.

**Comment by asubramanya** [comment]
@d Thanks for your suggestion. We have now replaced the gpos::HashValue with the `hashbpchar` and related functions as you suggested.

File: src/test/regress/expected/gporca_optimizer.out [hunk 6]
```diff
@@ -11111,9 +11111,355 @@ WITH abc AS (SELECT onetimefilter1.a, onetimefilter1.b FROM onetimefilter1, onet
...
+CREATE TABLE bpchar_table(a bpchar);
+NOTICE:  Table doesn't have 'DISTRIBUTED BY' clause -- Using column named 'a' as the Greenplum Database data distribution key for this table.
+HINT:  The 'DISTRIBUTED BY' clause determines the distribution of data. Make sure column(s) chosen are the optimal data distribution key to minimize skew.
+INSERT INTO bpchar_table SELECT 'a' || (i % 10) FROM generate_series(1, 1000) i;
+INSERT INTO bpchar_table SELECT 'a' || i FROM generate_series(1, 1000) i;
+ANALYZE bpchar_table;
+EXPLAIN SELECT * FROM bpchar_table WHERE a = 'a1';
+                                   QUERY PLAN                                   
+--------------------------------------------------------------------------------
+ Gather Motion 1:1  (slice1; segments: 1)  (cost=0.00..431.04 rows=101 width=3)
```
Comment (chrishajas) [inline src/test/regress/expected/gporca_optimizer.out]: I don't think gpdiff takes into account the cardinalities in the explain plan. If it does not, are these tests still providing value?

From atmsort.pl:
```

                                     QUERY PLAN                              
------------------------------------------------------------------------
 Gather Motion 2:1  (slice1)  (cost=0.00..698.88 rows=25088 width=550)
   Rows out:  150000 rows at destination with 0.230 ms to first row, 
   386 ms to end, start offset by 8.254 ms.
   ->  Seq Scan on customer  (cost=0.00..698.88 rows=25088 width=550)
         Rows out:  Avg 75000.0 rows x 2 workers.  Max 75001 rows (seg0) 
         with 0.056 ms to first row, 26 ms to end, start offset by 7.332 ms.
 Slice statistics:
   (slice0)    Executor memory: 186K bytes.
   (slice1)    Executor memory: 130K bytes avg x 2 workers, 
               130K bytes max (seg0).
 Total runtime: 413.401 ms
(8 rows)

atmsort reduces the plan to:

                                     QUERY PLAN                              
------------------------------------------------------------------------
{
  'child' => [
              {
      'id' => 2,
      'parent' => 1,
      'short' => 'Seq Scan on customer'
      }
  ],
  'id' => 1,
  'short' => 'Gather Motion'
  }
(8 rows)
```

Comment (asubramanya) [reply src/test/regress/expected/gporca_optimizer.out]: Yes. These tests may still be useful in catching crashes since they exercise the translator changes made as part of this PR, but maybe the existing tests already cover this. Either way its probably better to move these tests to ORCA.

**Comment by berlin-ab** [comment]
Are you ready for another round of feedback?

**Comment by asubramanya** [comment]
@berlin-ab Sure. Please let me know if you have additional feedback.

**Comment by asubramanya** [comment]
Pushed to master via https://github.com/greenplum-db/gpdb/commit/22d188898c02fc81c0bcba5142c9bd7c2e0650e3
